exposed/src/main/kotlin/org/jetbrains/exposed/sql/transactions/ThreadLocalTransactionManager.kt

```diff
@@ -7,138 +7,222 @@ import org.jetbrains.exposed.sql.Transaction
+            val currentManager = outer?.db.transactionManager
             try {
                 TransactionManager.resetCurrent(existingForDb)
-                it.statement()
+                transaction.statement().also {
+                    transaction.commit()
+                }
             } finally {
                 TransactionManager.resetCurrent(currentManager)
             }
-        } ?: inTopLevelTransaction(transactionIsolation, repetitionAttempts, db, statement)
+        } ?: inTopLevelTransaction(transactionIsolation, repetitionAttempts, db, null, statement)
     }
 }
 
-private fun TransactionInterface.rollbackLoggingException(log: (Exception) -> Unit){
+private fun TransactionInterface.rollbackLoggingException(log: (Exception) -> Unit) {
     try {
         rollback()
-    } catch (e: Exception){
+    } catch (e: Exception) {
         log(e)
     }
 }
 
-private inline fun TransactionInterface.closeLoggingException(log: (Exception) -> Unit){
+private inline fun TransactionInterface.closeLoggingException(log: (Exception) -> Unit) {
     try {
         close()
-    } catch (e: Exception){
+    } catch (e: Exception) {
         log(e)
     }
 }
 
-fun <T> inTopLevelTransaction(transactionIsolation: Int, repetitionAttempts: Int, db: Database? = null, statement: Transaction.() -> T): T {
-    var repetitions = 0
+fun <T> inTopLevelTransaction(
+    transactionIsolation: Int,
+    repetitionAttempts: Int,
+    db: Database? = null,
+    outerTransaction: Transaction? = null,
+    statement: Transaction.() -> T
+): T {
 
-    val outerManager = TransactionManager.manager.takeIf { TransactionManager.currentOrNull() != null }
-    while (true) {
-        db?.let { TransactionManager.managerFor(db)?.let { m -> TransactionManager.resetCurrent(m) } }
-        val transaction = TransactionManager.manager.newTransaction(transactionIsolation)
+    fun run():T {
+        var repetitions = 0
 
-        try {
-            val answer = transaction.statement()
-            transaction.commit()
-            return answer
-        }
-        catch (e: SQLException) {
-            val exposedSQLException = e as? ExposedSQLException
-            val queriesToLog = exposedSQLException?.causedByQueries()?.joinToString(";\n") ?: "${transaction.currentStatement}"
-            val message = "Transaction attempt #$repetitions failed: ${e.message}. Statement(s): $queriesToLog"
-            exposedSQLException?.contexts?.forEach {
-                transaction.interceptors.filterIsInstance<SqlLogger>().forEach { logger ->
-                    logger.log(it, transaction)
+        val outerManager = outerTransaction?.db.transactionManager.takeIf { it.currentOrNull() != null }
+
+        while (true) {
+            db?.let { db.transactionManager.let { m -> TransactionManager.resetCurrent(m) } }
+            val transaction = db.transactionManager.newTransaction(transactionIsolation, outerTransaction)
```

Comment:
Propagating `outerTransaction` will create nested transaction while inTopLevelTrasaction should always create new one.